### PR TITLE
New baselines for TagHelperBlockRewriterTest

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Part of #2584

Converted all of the TagHelperBlockRewriter tests. 
Only including the test changes. If curious, the source is here ([ParseTreeRewriter](https://github.com/aspnet/Razor/blob/ajbaaska/taghelper-parser/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperParseTreeRewriter.cs), [BlockRewriter](https://github.com/aspnet/Razor/blob/ajbaaska/taghelper-parser/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperBlockRewriter.cs)).

Here is the syntax for tag helpers,
``` XML
  <Node Name="MarkupTagHelperElementSyntax" Base="MarkupSyntaxNode">
    <Kind Name="MarkupTagHelperElement" />
    <Field Name="StartTag" Type="MarkupTagHelperStartTagSyntax" />
    <Field Name="Body" Type="SyntaxList&lt;RazorSyntaxNode&gt;" Optional="true" />
    <Field Name="EndTag" Type="MarkupTagHelperEndTagSyntax" Optional="true" />
  </Node>
  <Node Name="MarkupTagHelperStartTagSyntax" Base="RazorBlockSyntax">
    <Kind Name="MarkupTagHelperStartTag" />
    <Field Name="Children" Type="SyntaxList&lt;RazorSyntaxNode&gt;" Override="true" />
  </Node>
  <Node Name="MarkupTagHelperEndTagSyntax" Base="RazorBlockSyntax">
    <Kind Name="MarkupTagHelperEndTag" />
    <Field Name="Children" Type="SyntaxList&lt;RazorSyntaxNode&gt;" Override="true" />
  </Node>
  <Node Name="MarkupTagHelperAttributeSyntax" Base="MarkupSyntaxNode">
    <Kind Name="MarkupTagHelperAttribute" />
    <Field Name="NamePrefix" Type="MarkupTextLiteralSyntax" Optional="true" />
    <Field Name="Name" Type="MarkupTextLiteralSyntax" />
    <Field Name="NameSuffix" Type="MarkupTextLiteralSyntax" Optional="true" />
    <Field Name="EqualsToken" Type="SyntaxToken">
      <Kind Name="Equals" />
    </Field>
    <Field Name="ValuePrefix" Type="MarkupTextLiteralSyntax" Optional="true" />
    <Field Name="Value" Type="MarkupTagHelperAttributeValueSyntax" />
    <Field Name="ValueSuffix" Type="MarkupTextLiteralSyntax" Optional="true" />
  </Node>
  <Node Name="MarkupMinimizedTagHelperAttributeSyntax" Base="MarkupSyntaxNode">
    <Kind Name="MarkupMinimizedTagHelperAttribute" />
    <Field Name="NamePrefix" Type="MarkupTextLiteralSyntax" Optional="true" />
    <Field Name="Name" Type="MarkupTextLiteralSyntax" />
  </Node>
  <Node Name="MarkupTagHelperAttributeValueSyntax" Base="RazorBlockSyntax">
    <Kind Name="MarkupTagHelperAttributeValue" />
    <Field Name="Children" Type="SyntaxList&lt;RazorSyntaxNode&gt;" Override="true" />
  </Node>
```

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs [hunk 1]
```diff
@@ -69,9 +69,7 @@ private SyntaxNode RewriteNode(SyntaxNode node)
                     }
 
                     var tagName = tagBlock.GetTagName();
-                    if (string.IsNullOrWhiteSpace(tagName) ||
-                        tagBlock.IsVoidElement() ||
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs]: No longer special casing Void elements at this level (Just moved this logic to the tag helper phase).
We can always add the special casing later if we end up needing this for non-tag helper scenarios (Like Blazor).

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_CreatesErrorForIncompleteTagHelper2.stree.txt [hunk 2]
```diff
@@ -1,22 +1,36 @@
...
+                        CloseAngle;[>];
+                MarkupTextLiteral - [22..27)::5 - [World] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                    Text;[World];
+                MarkupTagHelperEndTag - [27..36)::9
+                    MarkupTextLiteral - [27..36)::9 - [</strong>] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                        OpenAngle;[<];
+                        ForwardSlash;[/];
+                        Text;[strong];
+                        CloseAngle;[>];
+        MarkupTagBlock - [36..42)::6 - [</div>]
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_CreatesErrorForIncompleteTagHelper2.stree.txt]: As discussed before, this `</div>` should not belong to the `p` tag. This was a bug. Hence the changes to the classified spans and tag helper spans.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_CreatesErrorForIncompleteTagHelper3.stree.txt [hunk 3]
```diff
@@ -1,22 +1,30 @@
...
+                Whitespace;[ ];
+            MarkupTagHelperElement - [14..27)::13 - strong[StartTagAndEndTag] - strongtaghelper
+                MarkupTagHelperStartTag - [14..22)::8
+                    MarkupTextLiteral - [14..22)::8 - [<strong>] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                        OpenAngle;[<];
+                        Text;[strong];
+                        CloseAngle;[>];
+                MarkupTextLiteral - [22..27)::5 - [World] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                    Text;[World];
+        MarkupTagBlock - [27..33)::6 - [</div>]
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_CreatesErrorForIncompleteTagHelper3.stree.txt]: > As discussed before, this </div> should not belong to the p tag. This was a bug. Hence the changes to the classified spans and tag helper spans.

Ditto

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CanHandleSymbolBoundAttributes1.diag.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+(0,0): Error RZ3003: Invalid tag helper bound property '[item]' on tag helper 'CatchAllTagHelper'. Tag helpers cannot bind to HTML attributes with name '[item]' because the name contains a '[' character.
+(0,0): Error RZ3003: Invalid tag helper bound property '[item]' on tag helper 'CatchAllTagHelper'. Tag helpers cannot bind to HTML attributes with name '[item]' because the name contains a ']' character.
+(0,0): Error RZ3003: Invalid tag helper bound property '[(item)]' on tag helper 'CatchAllTagHelper'. Tag helpers cannot bind to HTML attributes with name '[(item)]' because the name contains a '[' character.
+(0,0): Error RZ3003: Invalid tag helper bound property '[(item)]' on tag helper 'CatchAllTagHelper'. Tag helpers cannot bind to HTML attributes with name '[(item)]' because the name contains a ']' character.
+(0,0): Error RZ3003: Invalid tag helper bound property '*something' on tag helper 'CatchAllTagHelper'. Tag helpers cannot bind to HTML attributes with name '*something' because the name contains a '*' character.
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CanHandleSymbolBoundAttributes1.diag.txt]: > This is not a bug. These are the errors generated when the tag helper descriptors are found. Previously, TagHelperParseTreeRewriter was not responsible for combing these errors so test code never saw that. But I had to refactor some of that code and now the errors are combined in TagHelperParseTreeRewriter. Hence these errors show up.

Same reason as the [previous](https://github.com/aspnet/Razor/pull/2639#discussion_r223840033) PR.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/AllowsCompatibleTagStructures1.stree.txt [hunk 5]
```diff
@@ -1,3 +1,14 @@
...
+                MarkupTextLiteral - [0..7)::7 - [<input>] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                    OpenAngle;[<];
+                    Text;[input];
+                    CloseAngle;[>];
+            MarkupTagHelperEndTag - [7..15)::8
+                MarkupTextLiteral - [7..15)::8 - [</input>] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                    OpenAngle;[<];
+                    ForwardSlash;[/];
+                    Text;[input];
+                    CloseAngle;[>];
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/AllowsCompatibleTagStructures1.stree.txt]: Here is an example of why we want void elements to be grouped.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CreatesErrorForEmptyTagHelperBoundAttributes12.stree.txt [hunk 6]
```diff
@@ -1,8 +1,37 @@
...
+                    OpenAngle;[<];
+                    Text;[myth];
+                MarkupTagHelperAttribute - [5..12)::7 - BOUND - DoubleQuotes - [ BOUND=]
+                    MarkupTextLiteral - [5..6)::1 - [ ] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                        Whitespace;[ ];
+                    MarkupTextLiteral - [6..11)::5 - [BOUND] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                        Text;[BOUND];
+                    Equals;[=];
+                    MarkupTagHelperAttributeValue - [12..12)::0
+                        CSharpExpressionLiteral - [12..12)::0 - [] - Gen<None> - ImplicitExpressionEditHandler;Accepts:AnyExceptNewline;ImplicitExpression[ATD];K14
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CreatesErrorForEmptyTagHelperBoundAttributes12.stree.txt]: Fun fact: We want to generate a placeholder node when the taghelper attribute value is empty. We don't do that for non-taghelper attributes.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CreatesErrorForEmptyTagHelperBoundAttributes3.stree.txt [hunk 7]
```diff
@@ -1,6 +1,24 @@
...
+                MarkupTagHelperAttribute - [5..18)::13 - bound - SingleQuotes - [ bound='    ']
+                    MarkupTextLiteral - [5..6)::1 - [ ] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                        Whitespace;[ ];
+                    MarkupTextLiteral - [6..11)::5 - [bound] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                        Text;[bound];
+                    Equals;[=];
+                    MarkupTextLiteral - [12..13)::1 - ['] - Gen<None> - SpanEditHandler;Accepts:Any
+                        SingleQuote;['];
+                    MarkupTagHelperAttributeValue - [13..17)::4
+                        CSharpExpressionLiteral - [13..17)::4 - [    ] - Gen<None> - ImplicitExpressionEditHandler;Accepts:AnyExceptNewline;ImplicitExpression[ATD];K14
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CreatesErrorForEmptyTagHelperBoundAttributes3.stree.txt]: This is an example of what `TagHelperBlockRewriter` mainly does. It rewrites the attribute values for all non-string attributes to be code (aka `CSharpExpressionLiteral`). For a regular attribute, this whitespace will be represented as a `MarkupTextLiteral`

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/UnderstandsMinimizedAttributes_Document15.stree.txt [hunk 8]
```diff
@@ -1,10 +1,30 @@
...
-    Tag block - Gen<TagHelper> - 43 - (0:0,0) - input - InputTagHelper3
-        SelfClosing - <input string-prefix-value='some string' />
-        string-prefix-value - SingleQuotes
-            Markup block - Gen<None> - 11 - (28:0,28)
-                Markup span - Gen<Markup> - [some] - SpanEditHandler;Accepts:Any - (28:0,28) - Tokens:1
-                    SyntaxKind.Text;[some];
-                Markup span - Gen<Markup> - [ string] - SpanEditHandler;Accepts:Any - (32:0,32) - Tokens:2
-                    SyntaxKind.Whitespace;[ ];
-                    SyntaxKind.Text;[string];
+RazorDocument - [0..43)::43 - [<input string-prefix-value='some string' />]
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/UnderstandsMinimizedAttributes_Document15.stree.txt]: Not sure if this prefix is indicative of `@tagHelperPrefix` but it got me thinking; we should be representing `tagHelperPrefix` in the syntax tree of TagHelpers. Something for the cleanup issue.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_RewritesComplexAttributeTagHelperTagBlocks1.stree.txt [hunk 9]
```diff
@@ -1,23 +1,61 @@
...
-                            SyntaxKind.Dot;[.];
-                            SyntaxKind.Identifier;[Now];
+RazorDocument - [0..51)::51 - [<p class="@DateTime.Now" style='@DateTime.Now'></p>]
+    MarkupBlock - [0..51)::51
+        MarkupTagHelperElement - [0..51)::51 - p[StartTagAndEndTag] - ptaghelper
+            MarkupTagHelperStartTag - [0..47)::47
+                MarkupTextLiteral - [0..2)::2 - [<p] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                    OpenAngle;[<];
+                    Text;[p];
+                MarkupTagHelperAttribute - [2..24)::22 - class - DoubleQuotes - [ class="@DateTime.Now"]
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_RewritesComplexAttributeTagHelperTagBlocks1.stree.txt]: It'd be valuable in the baselines to understand which attributes are bound and which ones aren't.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/UnderstandsMinimizedAttributes_Document15.stree.txt (hunk 8 above)
Comment (ajaybhargavb) [reply test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/UnderstandsMinimizedAttributes_Document15.stree.txt]: Added

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_RewritesComplexAttributeTagHelperTagBlocks1.stree.txt (hunk 9 above)
Comment (ajaybhargavb) [reply test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/TagHelperParseTreeRewriter_RewritesComplexAttributeTagHelperTagBlocks1.stree.txt]: Good idea. Doing this now

**Comment by ajaybhargavb** [comment]
🆙 📅 . Baselines now show whether a tag helper attribute is bound or not.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CanHandleSymbolBoundAttributes1.stree.txt [hunk 10]
```diff
@@ -1,7 +1,35 @@
...
-            Code span - Gen<Expr> - [items] - ImplicitExpressionEditHandler;Accepts:AnyExceptNewline;ImplicitExpression[ATD];K14 - (18:0,18) - Tokens:1
-                SyntaxKind.Text;[items];
+RazorDocument - [0..30)::30 - [<ul bound [item]='items'></ul>]
+    MarkupBlock - [0..30)::30
+        MarkupTagHelperElement - [0..30)::30 - ul[StartTagAndEndTag] - CatchAllTagHelper
+            MarkupTagHelperStartTag - [0..25)::25
+                MarkupTextLiteral - [0..3)::3 - [<ul] - Gen<Markup> - SpanEditHandler;Accepts:Any
+                    OpenAngle;[<];
+                    Text;[ul];
+                MarkupMinimizedTagHelperAttribute - [3..9)::6 - [ bound]
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/ParserTests/TagHelperBlockRewriterTest/CanHandleSymbolBoundAttributes1.stree.txt]: Need same for minimized attributes.

**Comment by ajaybhargavb** [comment]
🆙 📅

**Pull request merged**